Split GoogleUtilities from FirebaseCore

paulb777 opened the pull request:
Create a new FirebaseUtilities pod consisting of Network, Environment, and Logger subspecs - extracted from FirebaseCore.  A follow-on PR will add a Swizzler subspec.

paulb777 commented:
This PR is not yet ready for detailed review. Current status is the six components in the Firebase Xcode project build and pass their unit tests with the new structure.  For now, I've left the Utilities unit tests combined with Core.

Next steps:
- Verify Firestore and Functions
- Verify static libraries
- Invent infrastructure to do do pod lib lint when FirebaseCore API's change
- Run copybara and integrate closed sourced dependencies

cc: @ryanwilson, @baolocdo, @tejasd

wilhuff commented:
Does this name (and grouping) at all align with Android? How does this relate to firebase_common over there?

It's getting hard to discuss/reason about dependencies across the platforms.

paulb777 commented:
@wilhuff I discussed with @bjornick and Android doesn't currently have shared utilities other than what might be in GMSCore.

paulb777 commented:
Travis is green and the PR is ready for review.  I'm planning to wait until after the 5.2.0 merge back to master before merging this one.

paulb777 commented:
@baolocdo Do you have a suggestion for something more specific than `Utiliities`? Utilities seems too non-descriptive for something being published in the CocoaPods namespace and something that will be used by many different Google products.  

What exactly is the reason for not mentioning Firebase? It seems that FirebaseUtilities is the most descriptive since the Firebase team owns the code and it the utilities are primarily used by Firebase. Analytics uses other utility pods already like GoogleToolboxForMac and nanopb.

paulb777 commented:
@salqadri Please help with the naming of the Utilities pod.

paulb777 commented:
Now ready for review:

- Split out GoogleUtilities from FirebaseCore
- Change prefix to GUL (for GoogleUtilitiesLibrary) for moved symbols
- Replace usage of GoogleToolboxForMac/NSData+zlib with a subspec in GoogleUtilities
- Add ALT_SOURCES and temporary tag mechanisms to enable pre-release lint testing when dependency pods change
- Keep FIRLogger as a wrapper around GULLogger with Firebase specific functionality and to make transition easier
- Split Reachability into its own subspec for FirebaseMessaging
- Replace GoogleToolboxForMac GTMDevLog usage in Messaging unit tests with NSLog
- Use Firebase/CoreOnly instead of Firebase/Core so that unit tests build and run without FirebaseAnalytics

Next step is integration with closed source dependencies

wilhuff commented on Firebase/Core/FIRLogger.m:
```diff
@@ -86,170 +64,53 @@
-    sMessageCodeRegex =
-        [NSRegularExpression regularExpressionWithPattern:kMessageCodePattern options:0 error:NULL];
-#endif
+    GULLoggerInitializeASL([arguments containsObject:kFIRDisableDebugModeApplicationArgument],
```
Was there an API review for this?

If not:
  * Let's drop "ASL", which is deprecated. We should eventually move to `os_log`.
  * Is there any way we can avoid passing three booleans to this? It's a terrible interface.
  * Is there a way to translate debug modes into a maximum log level instead so that GULLogger doesn't even have to be aware of this wart?

In particular it seems like if argument processing is to be done here, then we should completely figure out if we want debug mode from the arguments locally. Passing two booleans that seem like logical inverses of each other seems wrong.

ryanwilson replied:
+1 to moving to `os_log` but it's only available on iOS 10+, so we'll have to keep `ASL` around until we drop support for iOS 8 and 9 (will be a while).

wilhuff commented on Firebase/Core/FIRLogger.m:
```diff
@@ -86,170 +64,53 @@
+  sFIRAnalyticsDebugMode = analyticsDebugMode;
+  if (sFIRAnalyticsDebugMode) {
+    FIRLoggerInitializeASL();
+    sFIRAnalyticsDebugMode = analyticsDebugMode;
```
This seems to duplicate the assignment above.

Also, why make FIRLoggerInitializeASL conditional?

wilhuff commented on Firebase/Core/FIRLogger.m:
```diff
@@ -86,170 +64,53 @@
-void FIRSetLoggerLevel(FIRLoggerLevel loggerLevel) {
-  if (loggerLevel < FIRLoggerLevelMin || loggerLevel > FIRLoggerLevelMax) {
-    FIRLogError(kFIRLoggerCore, @"I-COR000023", @"Invalid logger level, %ld", (long)loggerLevel);
+  if (analyticsDebugMode && [GULAppEnvironmentUtil isFromAppStore]) {
```
This is modifying global state without a mutex or a queue. If this is intentional then you probably need a tsan annotation: `__attribute__((no_sanitize("thread")))`.

wilhuff commented on Firebase/Core/Private/FIRLogger.h:
```diff
@@ -79,6 +79,12 @@ void FIRSetLoggerLevel(FIRLoggerLevel loggerLevel);
 BOOL FIRIsLoggableLevel(FIRLoggerLevel loggerLevel, BOOL analyticsComponent);
 
+/**
+ * Register version to include in logs.
```
Which "version"? Is it the Firebase-wide version? The Core version?

Also, use `@param` to introduce the parameter description.

wilhuff commented on Firebase/Core/Public/FIRLoggerLevel.h:
```diff
@@ -14,20 +14,22 @@
  * limitations under the License.
  */
 
+#import <GoogleUtilities/GULLoggerLevel.h>
```
Now that GULLogger is nominally the public way to do this and FIRLogger is a compatibility shim for analytics, shouldn't we make FIRLoggerLevel a private header? I suppose that's a breaking change.

wilhuff commented on Firebase/Core/Public/FIRLoggerLevel.h:
```diff
@@ -14,20 +14,22 @@
 /**
  * The log levels used by internal logging.
  */
 typedef NS_ENUM(NSInteger, FIRLoggerLevel) {
```
Can't this whole thing now just be

```
typedef GULLoggerLevel FIRLoggerLevel;
```

wilhuff commented on Firebase/Utilities/Environment/third_party/GULAppEnvironmentUtil.m:
```diff
@@ -125,6 +125,24 @@ static BOOL isAppEncrypted() {
   return NO;
 }
 
+static BOOL hasSCInfoFolder() {
```
Free functions should start with a capital letter. Here and below.

http://google.github.io/styleguide/objcguide.html#function-names

wilhuff commented on Firebase/Utilities/Logger/GULLogger.m:
```diff
@@ -0,0 +1,273 @@
+#include <unistd.h>
+
+/// Arguments passed on launch.
+NSString *const kGULDisableDebugModeApplicationArgument = @"-GULDebugDisabled";
```
Unused.

wilhuff commented on Firebase/Utilities/Logger/GULLogger.m:
```diff
@@ -0,0 +1,273 @@
+NSString *const kGULLoggerForceSDTERRApplicationArgument = @"-GULLoggerForceSTDERR";
+
+/// Key for the debug mode bit in NSUserDefaults.
+NSString *const kGULPersistedDebugModeKey = @"/google/utilities/debug_mode";
```
This implementation is now split-brained. Why are we handling command-line arguments for debug mode in FIRLogger, but NSUserDefaults here?

Either insulate GULLogger from knowing about this gunk, or just put it all in here. This combined approach is even harder to reason about than the old system.

wilhuff commented on Firebase/Utilities/Logger/GULLogger.m:
```diff
@@ -0,0 +1,273 @@
+const char *kGULLoggerASLClientFacilityName = "com.google.utilities.app.logger";
+
+/// Keys for the number of errors and warnings logged.
+NSString *const kGULLoggerErrorCountKey = @"/google/utilities/count_of_errors_logged";
```
This is declared and defined but appears otherwise unused. Can we just drop it?

wilhuff commented on Firebase/Utilities/Logger/GULLogger.m:
```diff
@@ -0,0 +1,273 @@
+NSString *const kGULPersistedDebugModeKey = @"/google/utilities/debug_mode";
+
+/// ASL client facility name used by GULLogger.
+const char *kGULLoggerASLClientFacilityName = "com.google.utilities.app.logger";
```
This no longer really has anything to do with "app". "com.gogle.utilities.logger"?

wilhuff commented on Firebase/Utilities/Logger/GULLogger.m:
```diff
@@ -0,0 +1,273 @@
+}
+
+void GULSetAnalyticsDebugMode(BOOL analyticsDebugMode) {
+  GULLoggerInitializeASL(NO, NO, NO);
```
This is now functionally different than what came before.

Previously if your first call was FIRSetAnalyticsDebugMode we would initialize FIRLogger, *including evaluating the command-line arguments*. This now disregards the command-line arguments.

wilhuff commented:
At a high level I think we should try really hard to avoid putting any kind of "debug mode" in GULLogger. As implemented it's broken.

From my reading of the initialization code, debug mode has the effect of calling

```
asl_set_filter(sGULLoggerClient, ASL_FILTER_MASK_UPTO(ASL_LEVEL_DEBUG));
```

But it also sets `sGULAnalyticsDebugMode` and `sGULLoggerDebugMode`.

But GULLogBasic now has a "force" parameter and does not even concern itself with analytics debug mode.

GULIsLoggableLevel allows you to query by whether or not it's an analytics component but GULLogBasic doesn't. How is this supposed to work in practice?

I propose that analytics debug mode be pushed up into FIRLogger completely.

I'd also propose that we try to eliminate logger debug mode. Why isn't this equivalent to GULSetLoggerLevel to the max?

ryanwilson commented on Firebase/Utilities/Logger/GULLogger.m:
```diff
@@ -0,0 +1,273 @@
+// Copyright 2017 Google
```
I realize this was copied over, but does this count as new code and an updated copyright?

ryanwilson commented on Firebase/Auth/Source/FIRAuthAPNSTokenManager.m:
```diff
@@ -137,7 +137,7 @@ + (BOOL)isProductionApp {
   NSError *error = nil;
 
-  Class envClass = NSClassFromString(@"FIRAppEnvironmentUtil");
+  Class envClass = NSClassFromString(@"GULAppEnvironmentUtil");
```
Since Auth relies on `GoogleUtilities/GULAppEnvironmentUtil` already, can we remove the string based selectors here?

ryanwilson commented on FirebaseCore.podspec:
```diff
@@ -25,13 +27,12 @@ Firebase Core includes FIRApp and FIROptions which provide central configuration
   s.source_files = 'Firebase/Core/**/*.[mh]'
-  s.public_header_files = 'Firebase/Core/Public/*.h', 'Firebase/Core/Private/*.h'
-  s.private_header_files = 'Firebase/Core/Private/*.h', 'Firebase/Core/third_party/*.h'
+  s.public_header_files = 'Firebase/Core/Public/*.h'
```
What was the reason for having `Private/` here as well again?

ryanwilson commented on GoogleUtilities.podspec:
```diff
@@ -0,0 +1,65 @@
+  s.summary          = 'Firebase Utilities for iOS (plus community support for macOS and tvOS)'
+
+  s.description      = <<-DESC
+Firebase Utilities including Network, Environment, Logger, and Swizzling.
```
Do we want to have a note here saying not for public use, or "Internal Google Utilities for iOS including Network, Environment, Logger, and Swizzling"

ryanwilson commented on GoogleUtilities.podspec:
```diff
@@ -0,0 +1,65 @@
+Pod::Spec.new do |s|
+  s.name             = 'GoogleUtilities'
+  s.version          = '5.0.4'
+  s.summary          = 'Firebase Utilities for iOS (plus community support for macOS and tvOS)'
```
Nit: `s/Firebase/Google`

ryanwilson commented on Example/Firebase.xcodeproj/project.pbxproj:
```diff
@@ -4887,6 +4887,10 @@
 				MTL_ENABLE_DEBUG_INFO = YES;
+				OTHER_LDFLAGS = (
+					"$(inherited)",
+					"-ObjC",
```
Was this missed before, or why does it need to be added now?

paulb777 replied:
The new pod structure exposed macOS test failures from categories not being linked in. There's no change to iOS since their tests are linked with -all_load.

ryanwilson commented on Firebase/Core/FIRLogger.m:
```diff
@@ -14,16 +14,10 @@
-#import "FIRLoggerLevel.h"
+#import <FirebaseCore/FIRLoggerLevel.h>
+#import <GoogleUtilities/GULAppEnvironmentUtil.h>
+#import <GoogleUtilities/GULLogger.h>
```
Nit: Should have vertical whitespace between the framework imports and file import.

ryanwilson commented on Firebase/Core/FIRLogger.m:
```diff
@@ -86,97 +65,54 @@
-      aslOptions = 0;
-    }
-#endif  // TARGET_OS_SIMULATOR
+    // Register Firebase Version with GULLogger
```
Nit: punctuation.

ryanwilson commented on Firebase/Utilities/Network/Private/GULNetworkLoggerProtocol.h:
```diff
@@ -16,35 +16,35 @@
-  kFIRNetworkLogLevelInfo = FIRLoggerLevelInfo,
-  kFIRNetworkLogLevelDebug = FIRLoggerLevelDebug,
+#import <GoogleUtilities/GULLoggerLevel.h>
+#import "GULNetworkMessageCode.h"
```
Nit: Vertical whitespace between framework and file imports.

ryanwilson commented on GoogleUtilities.podspec:
```diff
@@ -0,0 +1,71 @@
+other Google CocoaPods. They're not intended for direct public usage.
+                       DESC
+
+  s.homepage         = 'https://firebase.google.com'
```
Should this be a more general homepage?

ryanwilson commented on GoogleUtilities.podspec:
```diff
@@ -0,0 +1,71 @@
+#    :tag => 'Utilities-' + s.version.to_s
+    :tag => 'pre-5.3-' + s.version.to_s
+  }
+  s.social_media_url = 'https://twitter.com/Firebase'
```
Same as above - maybe this should be `@googledevs`?

paulb777 commented:
@ryanwilson Thanks for the review! I'm going to merge now to unblock other tasks. I believe I've addressed most, if not all, of @wilhuff's requests. If there are more, I'll address them in a follow-up PR.

paulb777 commented:
Thanks @wilhuff. I'll make another PR to address these in the next few days. 

Like FIRLogger, the only part of the GULLogger API that is technically public is GULLoggerLevel.h. See https://github.com/firebase/firebase-ios-sdk/blob/master/GoogleUtilities.podspec#L38. However, the GoogleUtilities pod documents itself as to not supporting non-Google usage. We can incrementally improve everything else without impacting the public API. 

I don't think we should revert since this PR is an incremental step forward. It unblocks other projects that need to use GoogleUtilities. It provides size improvements for many Firebase configurations. And is a step towards cleaning up FIRLogger.

wilhuff commented:
Ah, interesting! I saw these listed in the public headers and thought this was a public API.

https://github.com/firebase/firebase-ios-sdk/blob/master/GoogleUtilities.podspec#L37

In that case please don't revert, but yes, let's address some of these rough edges as we get time.
